<a href="https://colab.research.google.com/github/kibedavid/Machine-Learning/blob/main/Exploratory_data_analysis_with_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Exploratory data analysis with Pandas 

Article outline
1. Demonstration of main Pandas methods
2. First attempt at predicting telecom churn

1. Demonstration of main Pandas methods

Pandas is a Python library that provides extensive means for data analysis. Data scientists often work with data stored in table formats like .csv, .tsv, or .xlsx. Pandas makes it very convenient to load, process, and analyze such tabular data using SQL-like queries. In conjunction with Matplotlib and Seaborn, Pandas provides a wide range of opportunities for visual analysis of tabular data

The main data structures in Pandas are implemented with Series and DataFrame classes. The former is a one-dimensional indexed array of some fixed data type. The latter is a two-dimensional data structure - a table - where each column contains data of the same type. You can see it as a dictionary of Series instances. DataFrames are great for representing real data: rows correspond to instances (examples, observations, etc.), and columns correspond to features of these instances.



In [3]:
import numpy as np
import pandas as pd
pd.set_option("display.precision", 2)

In [4]:
url = 'https://raw.githubusercontent.com/Yorko/mlcourse.ai/master/data/telecom_churn.csv'
df = pd.read_csv(url)
df.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1  False
1    OH             107  ...                       1  False
2    NJ             137  ...                       0  False
3    OH              84  ...                       2  False
4    OK              75  ...                       3  False

[5 rows x 20 columns]

Let’s have a look at data dimensionality, feature names, and feature types.

In [5]:
print(df.shape)

(3333, 20)


From the output, we can see that the table contains 3333 rows and 20 columns.

Now let's try printing out column names using columns:

In [6]:
print(df.columns)

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')


We can use the info() method to output some general information about the dataframe:

In [7]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

bool, int64, float64 and object are the data types of our features. We see that one feature is logical (bool), 3 features are of type object, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with shape.

We can change the column type with the astype method. Let's apply this method to the Churn feature to convert it into int64:

In [8]:
df['Churn'] = df['Churn'].astype('int64')

The describe method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

In [ ]:
df.describe()


Account length  Area code  ...  Customer service calls    Churn
count         3333.00    3333.00  ...                 3333.00  3333.00
mean           101.06     437.18  ...                    1.56     0.14
std             39.82      42.37  ...                    1.32     0.35
min              1.00     408.00  ...                    0.00     0.00
25%             74.00     408.00  ...                    1.00     0.00
50%            101.00     415.00  ...                    1.00     0.00
75%            127.00     510.00  ...                    2.00     0.00
max            243.00     510.00  ...                    9.00     1.00

[8 rows x 17 columns]

In order to see statistics on non-numerical features, one has to explicitly indicate data types of interest in the include parameter.

In [ ]:
df.describe(include=['object', 'bool'])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

For categorical (type object) and boolean (type bool) features we can use the value_counts method. Let's have a look at the distribution of Churn:

In [ ]:
df['Churn'].value_counts()

0    2850
1     483
Name: Churn, dtype: int64

2850 users out of 3333 are loyal; their Churn value is 0. To calculate fractions, pass normalize=True to the value_counts function.

In [ ]:
df['Churn'].value_counts(normalize=True)

0    0.86
1    0.14
Name: Churn, dtype: float64

Sorting


A DataFrame can be sorted by the value of one of the variables (i.e columns). For example, we can sort by Total day charge (use ascending=False to sort in descending order):

In [ ]:
df.sort_values(by='Total day charge', ascending=False).head()

State  Account length  ...  Customer service calls Churn
365     CO             154  ...                       1     1
985     NY              64  ...                       1     1
2594    OH             115  ...                       1     1
156     OH              83  ...                       0     1
605     MO             112  ...                       2     1

[5 rows x 20 columns]

We can also sort by multiple columns:

In [9]:
df.sort_values(by=['Churn', 'Total day charge'],
        ascending=[True, False]).head()

State  Account length  ...  Customer service calls Churn
688     MN              13  ...                       3     0
2259    NC             210  ...                       3     0
534     LA              67  ...                       4     0
575     SD             114  ...                       1     0
2858    AL             141  ...                       1     0

[5 rows x 20 columns]

**Indexing and retrieving data**


A DataFrame can be indexed in a few different ways.

To get a single column, you can use a DataFrame['Name'] construction. Let's use this to answer a question about that column alone: what is the proportion of churned users in our dataframe?

In [10]:
df['Churn'].mean()

0.14491449144914492

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt.

Boolean indexing with one column is also very convenient. The syntax is df[P(df['Name'])], where P is some logical condition that is checked for each element of the Name column. The result of such indexing is the DataFrame consisting only of rows that satisfy the P condition on the Name column.

Let's use it to answer the question:

**What are average values of numerical features for churned users?**

In [11]:
df[df['Churn'] == 1].mean()

Account length            102.66
Area code                 437.82
Number vmail messages       5.12
Total day minutes         206.91
Total day calls           101.34
Total day charge           35.18
Total eve minutes         212.41
Total eve calls           100.56
Total eve charge           18.05
Total night minutes       205.23
Total night calls         100.40
Total night charge          9.24
Total intl minutes         10.70
Total intl calls            4.16
Total intl charge           2.89
Customer service calls      2.23
Churn                       1.00
dtype: float64

**How much time (on average) do churned users spend on the phone during daytime?**

In [12]:
df[df['Churn'] == 1]['Total day minutes'].mean()

206.91407867494814

**What is the maximum length of international calls among loyal users (Churn == 0) who do not have an international plan?**



In [13]:
df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

DataFrames can be indexed by column name (label) or row name (index) or by the serial number of a row. The loc method is used for indexing by name, while iloc() is used for indexing by number.

In the first case below, we say "give us the values of the rows with index from 0 to 5 (inclusive) and columns labeled from State to Area code (inclusive)". In the second case, we say "give us the values of the first five rows in the first three columns" (as in a typical Python slice: the maximal value is not included).

In [14]:
df.loc[0:5, 'State':'Area code']

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [15]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

If we need the first or the last line of the data frame, we can use the df[:1] or df[-1:] construct:

In [16]:
df[-1:]

State  Account length  ...  Customer service calls Churn
3332    TN              74  ...                       0     0

[1 rows x 20 columns]

**Applying Functions to Cells, Columns and Rows**

To apply functions to each column, use apply():

In [17]:
df.apply(np.max) 

State                          WY
Account length                243
Area code                     510
International plan            Yes
Voice mail plan               Yes
Number vmail messages          51
Total day minutes         3.5e+02
Total day calls               165
Total day charge               60
Total eve minutes         3.6e+02
Total eve calls               170
Total eve charge               31
Total night minutes         4e+02
Total night calls             175
Total night charge             18
Total intl minutes             20
Total intl calls               20
Total intl charge             5.4
Customer service calls          9
Churn                           1
dtype: object


The apply method can also be used to apply a function to each row. To do this, specify axis=1. Lambda functions are very convenient in such scenarios. For example, if we need to select all states starting with W, we can do it like this:

In [18]:
df[df['State'].apply(lambda state: state[0] == 'W')].head()

State  Account length  ...  Customer service calls Churn
9     WV             141  ...                       0     0
26    WY              57  ...                       0     0
44    WI              64  ...                       1     0
49    WY              97  ...                       1     0
54    WY              87  ...                       5     1

[5 rows x 20 columns]

The map method can be used to replace values in a column by passing a dictionary of the form {old_value: new_value} as its argument:

In [19]:
d = {'No' : False, 'Yes' : True}
df['International plan'] = df['International plan'].map(d)
df.head()

State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1      0
1    OH             107  ...                       1      0
2    NJ             137  ...                       0      0
3    OH              84  ...                       2      0
4    OK              75  ...                       3      0

[5 rows x 20 columns]

The same thing can be done with the replace method:

In [20]:
df = df.replace({'Voice mail plan': d})
df.head()


State  Account length  ...  Customer service calls  Churn
0    KS             128  ...                       1      0
1    OH             107  ...                       1      0
2    NJ             137  ...                       0      0
3    OH              84  ...                       2      0
4    OK              75  ...                       3      0

[5 rows x 20 columns]